README.md
serve
=====

Windows | macOS | Linux
--------|-------|------
[![Build status](https://ci.appveyor.com/api/projects/status/nt16vsv7j1yk9yo2/branch/master?svg=true)](https://ci.appveyor.com/project/philippgille/serve/branch/master) | [![Build Status](https://travis-ci.org/philippgille/serve.svg?branch=master)](https://travis-ci.org/philippgille/serve/branches) | [![CircleCI](https://circleci.com/gh/philippgille/serve.svg?style=svg)](https://circleci.com/gh/philippgille/serve)

[![GitHub Releases](https://img.shields.io/github/release/philippgille/serve.svg)](https://github.com/philippgille/serve/releases)

`serve` starts a simple temporary static file server in your current directory and prints your IP address to share with colleagues.

It was based on [a Gist](https://gist.github.com/paulmach/7271283/2a1116ca15e34ee23ac5a3a87e2a626451424993) by [Paul Mach](https://github.com/paulmach), but has evolved a lot.

Contents
--------

- [Features](#features)
  - [Roadmap](#roadmap)
- [Install](#install)
    - [Windows](#windows)
    - [macOS](#macos)
    - [Linux](#linux)
    - [Manually](#manually)
- [Docker](#docker)
- [Use](#use)
    - [Example](#example)
- [Build](#build)
    - [Packages](#packages)
- [Related projects](#related-projects)

Features
--------

- [X] Serves static files in any given accessible directory until you hit `Ctrl-C`
- [X] Prints a table of your network interfaces and their IP addresses and guesses which one you most likely want to share
- [X] Follows softlinks transparently
- [X] Optional basic authentication
- [X] Optionally serve files via HTTPS instead of HTTP, using a temporary self signed certificate that's automatically generated for you on the fly
- [X] Optionally bind to a specific network interface, for example `localhost` to disable access from other devices
- [X] Only uses the Go standard library and no external dependencies
- [X] Single file executable, usable without any installation, or easy automatic updates when using a package manager

### Roadmap

- Find out and print external IP address that's reachable from the Internet
- Optionally compress files when requested
- Optionally make your server reachable from the Internet even if it's behind a router with NAT
  - Via reverse SSH tunnel
  - Via 3rd party service like [ngrok](https://ngrok.com/), [https://localtunnel.me](https://localtunnel.me), [LabStack Tunnel](https://labstack.com/docs/tunnel) etc.
  - Via [frp](https://github.com/fatedier/frp)
- Optionally run a [Tor](https://www.torproject.org/) hidden service (v3) for:
  - Automatic accessibility from the Internet (via Tor Browser or proxy) even when behind a router with NAT
  - Encrypted traffic
  - No exposure of your IP address

Install
-------

We recommend installing `serve` with one of the following package managers, because they provide you with functionality such as automatic updates, instant availability as command in the `PATH`, easy removal, sandboxing (depending on the package manager) etc.  
But alternatively you can always install `serve` manually as well, see [Manually](#manually).

### Windows

The easiest way is to use the package manager [Scoop](http://scoop.sh/):  
`scoop install serve`

Another option is [Chocolatey](https://chocolatey.org/):  
`choco install serve`

You can also have a look at the description in the Chocolatey Gallery on [https://chocolatey.org/packages/serve/](https://chocolatey.org/packages/serve/).

### macOS

The easiest way is to use the package manager [Homebrew](https://brew.sh/):  
`brew tap philippgille/tap`  
`brew install serve`

Or in a single command:  
`brew install philippgille/tap/serve`

### Linux

The easiest way is to use the package manager [Snap](https://snapcraft.io/), which is installed by default on Ubuntu 16.04 and later:  
`snap install serve`

You can also have a look at the description in the Snap Store on [https://snapcraft.io/serve](https://snapcraft.io/serve).

> Note: Due to sandboxing by Snap, `serve` can only serve files in the user's `$HOME` directory.

### Manually

#### With Go installed

`go get -u github.com/philippgille/serve`

> Note: Requires your `$GOPATH/bin` directory to be in your `PATH`, which is usually the case.

#### Without Go installed

You can download the binary for your OS from the [releases](https://github.com/philippgille/serve/releases) simply make it available as command in your `PATH`. See [Manual Installation](https://github.com/philippgille/serve/tree/master/docs#manual-installation) for details.

Docker
------

`serve` is also available as Docker image in the Docker Hub: [https://hub.docker.com/r/philippgille/serve/](https://hub.docker.com/r/philippgille/serve/)

Please read [docker/README.md](https://github.com/philippgille/serve/blob/master/docker/README.md) for information on how to use it.

Use
---

```
$ serve -h
Usage of serve:
  -a string
        Require basic authentication with the given credentials (e.g. -a "alice:secret")
  -b string
        Bind to (listen on) a specific interface. "0.0.0.0" is for ALL interfaces. "localhost" disables access from other devices. (default "0.0.0.0")
  -d string
        The directory of static files to host (default ".")
  -h    Print the usage
  -p string
        Port to serve on. 8080 by default for HTTP, 8443 for HTTPS (when using the -s flag) (default "8080")
  -s    Serve via HTTPS instead of HTTP. Creates a temporary self-signed certificate for localhost, 127.0.0.1, <hostname>.local, <hostname>.lan, <hostname>.home and the determined LAN IP address
  -t    Test / dry run (just prints the interface table)
  -v    Print the version
```

Press `Ctrl+C` in the terminal to stop the server.

### Example

```bash
~/path/to/servable/files$ serve

Serving "." on all network interfaces (0.0.0.0) on HTTP port: 8080

Local network interfaces and their IP address so you can pass one to your colleagues:

      Interface      |  IPv4 Address   |              IPv6 Address
---------------------|-----------------|----------------------------------------
lo                   | 127.0.0.1       | ::1
eth0                 |                 | 
wlan0                | 192.168.178.123 | fe80::e7b:fdaf:ae5d:3cfa
virbr0               | 192.168.122.1   | 
br-8ef347e8a4e9      | 172.22.0.1      | fe80::42:c9ff:fed3:35a
docker_gwbridge      | 172.21.0.1      | 
docker0              | 172.17.0.1      | fe80::42:c6cf:fe3d:a554
veth0d522f4          |                 | fe80::307a:7fcf:fe3d:cba4

You probably want to share:
http://192.168.178.123:8080
```

When opening the URL `http://192.168.178.123:8080` in a browser you see the directory you're serving. For example:

![screenshot](assets/example-2-browser.png)

Build
-----

To build `serve` by yourself:

1. [Install Go](https://golang.org/doc/install)
2. `cd` into the root directory of this repository
3. Execute: `go build`

> Note: The binaries in GitHub Releases are shrinked with additional Go linker flags and UPX

To also make `serve` available as command in other directories:

1. Add `$GOPATH/bin` to your `PATH` if you haven't done that already when installing Go
2. Execute: `go install`

There are also build scripts for Windows and Linux for creating release artifacts (shrinked binaries for Windows, macOS and Linux):

- Windows: `build/build.ps1`
- Linux: `build/build.sh`

> Note: They require Go and UPX to be installed

To build with a Docker container:

`docker run --rm -v $(pwd):/go/src/github.com/philippgille/serve -w /go/src/github.com/philippgille/serve golang build/build.sh noupx`  
Or with UPX:  
`docker run --rm -v $(pwd):/go/src/github.com/philippgille/serve -w /go/src/github.com/philippgille/serve golang bash -c "apt update && apt install -y upx-ucl && build/build.sh"`

> Note: You have to use `${pwd}` instead of `$(pwd)` on Windows.

### Packages

For Scoop and Homebrew no packages need to be built. They use "manifests"/"formulae" and the binaries from GitHub Releases.

For releasing a new version, they need to be updated here:

- Scoop: [https://github.com/lukesampson/scoop/blob/master/bucket/serve.json](https://github.com/lukesampson/scoop/blob/master/bucket/serve.json)
- Homebrew: [https://github.com/philippgille/homebrew-tap/blob/master/serve.rb](https://github.com/philippgille/homebrew-tap/blob/master/serve.rb)

For Snap a Git hook is set up in the Snapcraft dashboard to automatically build a new Snap on every commit, so for releasing a new version the file in *this* repository needs to be updated:

- Snap: [https://github.com/philippgille/serve/blob/master/snap/snapcraft.yaml](https://github.com/philippgille/serve/blob/master/snap/snapcraft.yaml)

The Snap package can also be built manually. In the past this could even be done within a Docker container, but the [official Snapcraft Docker image](https://hub.docker.com/u/snapcoret/snapcraft) (according to the [docs](https://docs.snapcraft.io/build-on-docker/4158)) is outdated (as of 2019-05-01) and doesn't contain the latest version of `snapcraft` (and installing the latest version via `snap` itself, as you'd do nowadays according to the official docs, doesn't work).  
So now you can only build the Snap package on Linux, using the following steps:

1. `snap install snapcraft --classic`
2. `snapcraft`

Depending on the current `serve` version and your CPU's architecture it will create a file like `serve_0.3.2_amd64.snap`, which can manually be installed with `snap install --dangerous serve_0.3.2_amd64.snap`.

The Chocolatey packages need to be uploaded manually to Chocolatey [here](https://chocolatey.org/packages/upload). The package can be built with this script:

- Windows: `build\build-chocolatey.ps1`

The Docker image can be built like this:

`docker build -f docker/Dockerfile -t philippgille/serve .`

Related projects
----------------

- `python -m SimpleHTTPServer 8080`
  - Con: Requires Python, no option to require authentication, no HTTPS
- [https://github.com/indexzero/http-server](https://github.com/indexzero/http-server)
  - Pro: Popular (8200 GitHub stars as of 2019-05-02), mature, feature-rich
  - Con: Requires Node.js
- [https://github.com/zeit/serve](https://github.com/zeit/serve)
  - Pro: Popular (4200 GitHub stars as of 2019-05-02), mature
  - Con: Requires Node.js, no option to require authentication, no HTTPS, can't serve directories other than the current working directory
- [https://github.com/codeskyblue/gohttpserver](https://github.com/codeskyblue/gohttpserver)
  - Pro: Nice web UI, shows QR codes for downloading files from a smartphone, OpenID and OAuth2 authentication, optional upload of files from a client, `README.md` preview, directory zip download any many more features
  - Con: Too many features ("[feature creep](https://en.wikipedia.org/wiki/Feature_creep)")? Many dependencies.
- [https://github.com/syntaqx/serve](https://github.com/syntaqx/serve)
  - Con: No option to require authentication, no installation packages for Windows or Linux
- [https://github.com/rhardih/serve](https://github.com/rhardih/serve)
  - Con: No option to require authentication, no installation packages, no HTTPS when _not_ using HTTP/2, when using HTTP/2 the certificate and private key are written to disk in the current working directory, *which is served by default*, so an attacker can easily download and use them in a Man-in-the-Middle attack without the client noticing (because it's the correct certificate) (as of 2019-05-05 - I created [an issue](https://github.com/rhardih/serve/issues/3) for that and hope it gets fixed soon)
- Many others!


RELEASES.md
Releases
========

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

vNext
-----

v0.3.2 (2019-07-20)
-------------------

> Note: Windows and Linux users who have `serve` v0.3.0 don't need to update to this version, because it only fixes a bug in v0.3.1 and v0.3.1 only contained a bugfix for macOS.

- Fixed: serve v0.3.1 reports version v0.3.0 when using `serve -v`

v0.3.1 (2019-07-20)
-------------------

> Note: This release only contains a bugfix for macOS. There's no need to update from v0.3.0 to this version on Windows or Linux.

- Fixed: `serve` crashes at start on macOS due to a bug in UPX, which is used to compress the executable file ([issue #27](https://github.com/philippgille/serve/issues/27))
  - See [https://github.com/upx/upx/issues/222](https://github.com/upx/upx/issues/222)
  - Only macOS is affected!

v0.3.0 (2019-05-04)
-------------------

- Added: Optional generation and use of a self signed certificate to serve files via HTTPS instead of HTTP ([issue #9](https://github.com/philippgille/serve/issues/9))
- Added: Optional basic authentication ([issue #10](https://github.com/philippgille/serve/issues/10))
- Added: Option to bind to a specific network interface ([issue #19](https://github.com/philippgille/serve/issues/19))
- Added: Exit with an error when using the `-d` flag and the argument is not a directory, the directory doesn't exist or the directory is not readable
- Added: Handling of the directory as positional argument. We keep this undocumented for now to not promote its use, but it was a common source of problems (where a directory was passed as positional argument and then "." was served), which are now mitigated.
- Improved: Updated building the snap with snapcraft 3.x instead of its 2.4x legacy mode
- Fixed: `serve -h` lead to exit code 2 instead of 0
- Fixed: Wrong documentation URL in the Chocolatey Nuspec file ([issue #18](https://github.com/philippgille/serve/issues/18))

### Breaking changes:

- Changed default port from 8100 to 8080, which is an official alternative to port 80 for HTTP
- Docker:
  - Changed default serving directory within the container from `/share` to `/srv`, which is a common directory for serving files
  - Moved `-d` flag from `CMD` to `ENTRYPOINT`, leading to fewer problems when using the Docker container with additional flags

v0.2.1 (2018-05-13)
-------------------

- Improved: Increased interface table width to 80 for long interface names
- Fixed: No suggested URL in several cases ([issue #7](https://github.com/philippgille/serve/issues/7))
- Fixed: Snap package doesn't work ([issue #14](https://github.com/philippgille/serve/issues/14))

v0.2.0 (2018-05-01)
-------------------

- Added: Version flag to print the version of the CLI
- Added: Test flag for a dry run which only prints the network interface table
- Improved: Released binaries are much smaller because they get shrinked with UPX ([issue #1](https://github.com/philippgille/serve/issues/1))
- Fixed: Network interface table looks bad on Windows ([issue #2](https://github.com/philippgille/serve/issues/2))
- Fixed: IPv4 address doesn't get printed for the main network interface on Windows ([issue #3](https://github.com/philippgille/serve/issues/3))

v0.1.0 (2018-05-01)
-------------------

- Added: Basic temporary static file server that prints the local interfaces and their addresses


net.go
package main

import (
	"log"
	"net"
	"os"
	"runtime"
	"strings"
)

// getAddressesFromIface goes through the addresses of the given interface and tries to return the first of each kind.
//
// The interesting interfaces like eth0 and wlan0 typically have 2 addresses: one IPv4 and one IPv6 address.
// But some interfaces just have one of them, or if an interface is deactivated it doesn't have any.
// On Windows the main network interface like "Ethernet 3" can have many addresses and the main IPv4 address doesn't have to be one of the first 2.
// We must take care of all these combinations.
func getAddressesFromIface(iface net.Interface) (ipv4 string, ipv6 string) {
	addrs, err := iface.Addrs()
	if err != nil {
		log.Fatal(err)
	}
	for i := 0; i < len(addrs) && (ipv4 == "" || ipv6 == ""); i++ {
		// In the case of two addresses they could potentially be of the same type.
		// We want to show the first address. overwriteIfEmpty() doesn't overwrite existing values.
		addrWithoutMask := strings.Split(addrs[i].String(), "/")[0]
		if strings.Contains(addrWithoutMask, ":") {
			overwriteIfEmpty(&ipv4, "")
			overwriteIfEmpty(&ipv6, addrWithoutMask)
		} else {
			overwriteIfEmpty(&ipv4, addrWithoutMask)
			overwriteIfEmpty(&ipv6, "")
		}
	}
	return
}

// isFav checks the network interface's name and if it's a typical main one (like "eth0" on Linux) it returns true.
//
// Note: All possible runtime.GOOS values are listed here: https://golang.org/doc/install/source#environment
func isFav(iface net.Interface) bool {
	switch runtime.GOOS {
	case "windows":
		if iface.Name == "WiFi" ||
			len(iface.Name) >= 8 && iface.Name[:8] == "Ethernet" {
			return true
		}
	case "darwin":
		if iface.Name == "en0" || iface.Name == "en1" {
			return true
		}
	case "linux":
		if iface.Name == "eth0" || iface.Name == "wlan0" {
			return true
		}
	}
	return false
}

// defaultSANs returns DNS names and IP addresses that might be used to reach the current host,
// either from the host itself or from other machines in the local network.
func defaultSANs() []string {
	result := []string{"localhost", "127.0.0.1"}

	hostname, err := os.Hostname()
	if err == nil {
		result = append(result, hostname, hostname+".local", "*."+hostname+".local", hostname+".lan", "*."+hostname+".lan", hostname+".home", "*."+hostname+".home")
	}

	lanIP, err := lanIP()
	if err == nil {
		result = append(result, lanIP)
	}

	return result
}

// lanIP tries to determine the IP address of the current machine in the LAN.
func lanIP() (string, error) {
	ifaces, err := net.Interfaces()
	if err != nil {
		return "", err
	}
	fav := ""
	for _, iface := range ifaces {
		if isFav(iface) {
			fav, _ = getAddressesFromIface(iface)
			break
		}
	}
	return fav, nil
}


tls.go
package main

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/tls"
	"crypto/x509"
	"crypto/x509/pkix"
	"fmt"
	"math"
	"math/big"
	"net"
	"strings"
	"time"
)

const certValidityDuration time.Duration = 24 * time.Hour * 7 // 7 days
var certValidityDurationDays int = int(math.Floor(certValidityDuration.Hours()) / 24)

// generateCert generates a self-signed certificate for the given and some default SANs.
// It's based on
// https://github.com/mholt/caddy/blob/master/caddytls/selfsigned.go and
// https://github.com/gerald1248/httpscerts/tree/diskless (which is based on
// https://github.com/kabukky/httpscerts, which is based on
// https://golang.org/src/crypto/tls/generate_cert.go).
func generateCert() (tls.Certificate, []string, error) {
	privKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return tls.Certificate{}, nil, fmt.Errorf("Failed to generate private key: %v", err)
	}

	notBefore := time.Now()
	notAfter := notBefore.Add(certValidityDuration)

	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
	if err != nil {
		return tls.Certificate{}, nil, fmt.Errorf("Failed to generate serial number: %v", err)
	}

	cert := x509.Certificate{
		SerialNumber: serialNumber,
		Subject:      pkix.Name{Organization: []string{"serve self-signed"}},
		NotBefore:    notBefore,
		NotAfter:     notAfter,
		KeyUsage:     x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
		ExtKeyUsage:  []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
		// The certificate is used directly and not as CA to sign other certificates,
		// so we don't need `IsCA: true` or `KeyUsage |= x509.KeyUsageCertSign`.
		// We also don't need `BasicConstraintsValid: true`,
		// because this certificate doesn't need any validation, because it's created here.
	}

	// SAN = Subject Alternative Name / "subjectAltName".
	// The CN ("Common Name") seems to be deprecated in favor of SAN.
	// Usually used for hostname validation by clients.
	//
	// Note: Neither Firefox nor Chrome seem to validate the hostname for a self signed certificate
	// when its CA is not installed on the host. Instead they show the warning regarding the untrusted CA,
	// and when temporarily trusted within the browser no further warning is shown.
	// Microsoft Edge does it the other way around: When the hostname doesn't match a SAN entry it shows the appropriate warning,
	// but no warning regarding the untrusted CA. This is only shown when the hostname matches the SAN entry.
	//
	// However, in case the client installs the certificate on its host, or if some client does both checks,
	// having some SAN entries that might match the hostname might make sense.
	sans := defaultSANs()
	for _, san := range sans {
		if ip := net.ParseIP(san); ip != nil {
			cert.IPAddresses = append(cert.IPAddresses, ip)
		} else {
			cert.DNSNames = append(cert.DNSNames, strings.ToLower(san))
		}
	}

	derBytes, err := x509.CreateCertificate(rand.Reader, &cert, &cert, &privKey.PublicKey, privKey)
	if err != nil {
		return tls.Certificate{}, nil, fmt.Errorf("Could not create certificate: %v", err)
	}

	chain := [][]byte{derBytes}

	return tls.Certificate{
		Certificate: chain,
		PrivateKey:  privKey,
		Leaf:        &cert,
	}, sans, nil
}


util.go
package main

import (
	"flag"
	"log"
	"os"
)

// cutString cuts strings that exceed the maxLen to (maxLen-2) and adds ".."
func cutString(s string, maxLen int) string {
	if len(s) > maxLen {
		return s[:maxLen-2] + ".."
	}
	return s
}

// overwriteIfEmpty only overwrites the string s with the string overwrite if s is empty
func overwriteIfEmpty(s *string, overwrite string) {
	if *s == "" {
		*s = overwrite
	}
}

// isFlagPassed returns true if the flag was explicitly set as CLI parameter
func isFlagPassed(name string) bool {
	found := false
	flag.Visit(func(f *flag.Flag) {
		if f.Name == name {
			found = true
		}
	})
	return found
}

// isDirAccessible checks if the directory exsists, if it really is a directory instead of a file and if it's readable.
// If logFatal is true, errors will be printed and the program will exit. If it's false the program will just return the result.
func isDirAccessible(dir string, logFatal bool) bool {
	// Use Stat instead of Lstat because serving works with softlinks (e.g. "serve -d ./softlink")
	fileInfo, err := os.Stat(*directory)
	if err != nil {
		if logFatal {
			log.Fatalf("%v can't be served: %v\n", *directory, err)
		}
		return false
	} else if !fileInfo.IsDir() {
		if logFatal {
			log.Fatalf("%v can't be served because it's a file and not a directory: %v\n", *directory, err)
		}
		return false
	} else {
		file, err := os.Open(*directory)
		if err != nil {
			if logFatal {
				log.Fatalf("%v can't be served because it's not readable: %v\n", *directory, err)
			}
			return false
		} else {
			file.Close() // Ignore errors
		}
	}
	return true
}


build.sh
#!/bin/bash

# Script for building the serve binaries.
# 6 artifacts are built: A binary for Windows, macOS and Linux (all x64) and an archive of each.
#
# Example: "./build.sh"
# Example for building without UPX compression: "./build.sh noupx"

set -euxo pipefail

# Parameters
if [[ $# -eq 1 && "$1" == "noupx" ]]; then
    NO_UPX=true
else
    NO_UPX=false
fi

SCRIPTDIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

ARTIFACTSDIR="$SCRIPTDIR/artifacts"
VERSION=$(<$SCRIPTDIR/VERSION)

# Clean up the previous build
rm -rf ${ARTIFACTSDIR}

# Build for Windows, macOS and Linux
# Use linker flags for shrinking
GOOS=windows GOARCH=amd64 go build -v -o "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64/serve.exe" -ldflags="-s -w" "github.com/philippgille/serve"
GOOS=darwin GOARCH=amd64 go build -v -o "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64/serve" -ldflags="-s -w" "github.com/philippgille/serve"
GOOS=linux GOARCH=amd64 go build -v -o "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64/serve" -ldflags="-s -w" "github.com/philippgille/serve"

# Shrink binaries with UPX.
# Requires UPX to be installed (for example with "apt install upx-ucl").
if [[ "$NO_UPX" == "false" ]]; then
    upx --ultra-brute "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64/serve.exe"
    upx --ultra-brute "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64/serve"
    upx --ultra-brute "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64/serve"
fi

# Create an archive for each of the "serve" binaries, so when users extract the archive, they don't have to rename it
declare -a arr=("Windows" "macOS" "Linux")
for MYOS in "${arr[@]}"
do
    # Sleep to prevent: tar: serve_v0.2.0_macOS_x64: file changed as we read it
    sleep 1s
    tar -czf "${ARTIFACTSDIR}/serve_v${VERSION}_${MYOS}_x64.tar.gz" -C "${ARTIFACTSDIR}" "serve_v${VERSION}_${MYOS}_x64"
done

# Also copy and rename the original files to have bare binaries
cp "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64/serve.exe" "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64.exe"
rm -rf "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64"
cp "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64/serve" "${ARTIFACTSDIR}/serve"
rm -rf "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64"
mv "${ARTIFACTSDIR}/serve" "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64"
cp "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64/serve" "${ARTIFACTSDIR}/serve"
rm -rf "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64"
mv "${ARTIFACTSDIR}/serve" "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64"

# Generate hashes and save as file.
# Archives
echo -n $(sha256sum "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64.tar.gz" | cut -d " " -f 1) > "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64.tar.gz.sha256"
echo -n $(sha256sum "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64.tar.gz" | cut -d " " -f 1) > "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64.tar.gz.sha256"
echo -n $(sha256sum "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64.tar.gz" | cut -d " " -f 1) > "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64.tar.gz.sha256"
# Binaries
echo -n $(sha256sum "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64.exe" | cut -d " " -f 1) > "${ARTIFACTSDIR}/serve_v${VERSION}_Windows_x64.exe.sha256"
echo -n $(sha256sum "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64" | cut -d " " -f 1) > "${ARTIFACTSDIR}/serve_v${VERSION}_macOS_x64.sha256"
echo -n $(sha256sum "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64" | cut -d " " -f 1) > "${ARTIFACTSDIR}/serve_v${VERSION}_Linux_x64.sha256"


serve.go
/*
Serve starts a simple temporary static file server in your current directory and prints your IP address to share with colleagues.

Initially based on the Gist https://gist.github.com/paulmach/7271283/2a1116ca15e34ee23ac5a3a87e2a626451424993
by Paul Mach (https://github.com/paulmach).

Usage:
  -a string
        Require basic authentication with the given credentials (e.g. -a "alice:secret")
  -b string
        Bind to (listen on) a specific interface. "0.0.0.0" is for ALL interfaces. "localhost" disables access from other devices. (default "0.0.0.0")
  -d string
        The directory of static files to host (default ".")
  -h    Print the usage
  -p string
        Port to serve on. 8080 by default for HTTP, 8443 for HTTPS (when using the -s flag) (default "8080")
  -s    Serve via HTTPS instead of HTTP. Creates a temporary self-signed certificate for localhost, 127.0.0.1, <hostname>.local, <hostname>.lan, <hostname>.home and the determined LAN IP address
  -t    Test / dry run (just prints the interface table)
  -v    Print the version

Navigating to http://localhost:8080 will display the index.html or directory listing file.
*/
package main

import (
	"crypto/sha1"
	"crypto/tls"
	"flag"
	"fmt"
	"log"
	"net"
	"net/http"
	"os"
	"strings"
)

// Increment and remove "+" in release commits.
// Add "+" after release commits.
const version = "v0.3.2"

// Flags in alphabetical order, just like "-h" prints them
var (
	auth         = flag.String("a", "", `Require basic authentication with the given credentials (e.g. -a "alice:secret")`)
	bind         = flag.String("b", "0.0.0.0", `Bind to (listen on) a specific interface. "0.0.0.0" is for ALL interfaces. "localhost" disables access from other devices.`)
	directory    = flag.String("d", ".", "The directory of static files to host")
	help         = flag.Bool("h", false, "Print the usage")
	port         = flag.String("p", "8080", "Port to serve on. 8080 by default for HTTP, 8443 for HTTPS (when using the -s flag)")
	https        = flag.Bool("s", false, "Serve via HTTPS instead of HTTP. Creates a temporary self-signed certificate for localhost, 127.0.0.1, <hostname>.local, <hostname>.lan, <hostname>.home and the determined LAN IP address")
	test         = flag.Bool("t", false, "Test / dry run (just prints the interface table)")
	printVersion = flag.Bool("v", false, "Print the version")
)

func main() {
	flag.Parse()

	// If the "-h" flag was used, only print the usage and exit.
	// Without the "-h" flag in the flag definitions "-h" would also print the usage,
	// but only because it's an unknown flag and then exit with exit code 2 (instead of 0).
	if *help {
		// flag.Usage() prints to stderr by default because it's usually called internally
		// (in the flag package) when an invalid argument is used.
		// But in our case it's meant to be printed, so print to stdout.
		flag.CommandLine.SetOutput(os.Stdout)
		flag.Usage()
		os.Exit(0)
	}

	// If the "-v" flag was used, only print the version and exit
	if *printVersion {
		fmt.Printf("serve version: %v\n", version)
		os.Exit(0)
	}

	// If the "-d" flag wasn't used, but a POSITIONAL argument is passed,
	// this likely means the user wants to serve that directory,
	// because many other CLIs expect the main subject as first positional argument.
	// Examples: `go build -v .`; `docker run -it busybox`; `git add -v *.go`
	if !isFlagPassed("d") && flag.Arg(0) != "" && isDirAccessible(flag.Arg(0), false) {
		fmt.Printf("\nWARNING: You didn't use the \"-d\" flag, but a positional argument instead. It seems to be a valid directory though, so serve is going to serve it.\n")
		*directory = flag.Arg(0)
	}

	// Precondition checks.
	// Should be done for most string flags.

	// Auth
	if *auth != "" && !strings.Contains(*auth, ":") {
		log.Fatal(`When using the -a flag to add basic authentication, you must specify credentials in the form of "username:password". For example: "alice:secret".`)
	}
	// Bind: Will be detected by a lookup when calling ListenAndServe() at the end
	// Directory
	// Note: All checks in the following call aren't necessary, but user-friendly.
	// Not necessary because the dir must only exist and be a dir and be readable on each request.
	// So theoretically the server could already be started and the directory created/modified later.
	isDirAccessible(*directory, true)
	// Port: Will be detected by a lookup when calling ListenAndServe() at the end

	scheme := "http"
	if *https {
		scheme += "s"
		// If the port wasn't specifically set by the user,
		// change it from its default value 8080 to 8443 as mentioned in the flag description.
		if !isFlagPassed("p") {
			*port = "8443"
		}
	}

	// If the "-t" flag was used, only print the network interface table and exit
	if *test {
		if *bind != "0.0.0.0" {
			fmt.Printf("\nNo need to print the network interface table, because serve will bind to %v as requested, making it reachable via:\n%v://%v:%v\n", *bind, scheme, *bind, *port)
		} else {
			printAddrs(scheme, *port)
		}
		os.Exit(0)
	}

	finalHandler := http.FileServer(http.Dir(*directory))

	// If the "-a" flag was used, use basic authentication middleware
	if *auth != "" {
		finalHandler = withBasicAuth(finalHandler)
	}

	// Register handler for "/" in Go's DefaultServeMux
	http.Handle("/", finalHandler)

	if *bind == "0.0.0.0" {
		fmt.Printf("\nServing \"%s\" on all network interfaces (0.0.0.0) on %v port: %s\n", *directory, strings.ToUpper(scheme), *port)
		// Print local network interfaces and their IP addresses
		printAddrs(scheme, *port)
	} else {
		fmt.Printf("\nServing \"%s\" on:\n%v://%v:%s\n", *directory, scheme, *bind, *port)
	}

	if *https {
		cert, sans, err := generateCert()
		if err != nil {
			log.Fatalf("Couldn't generate TLS certificates: %v\n", err)
		}
		tlsConfig := &tls.Config{
			MinVersion:               tls.VersionTLS12,
			PreferServerCipherSuites: true,
			Certificates:             []tls.Certificate{cert},
		}
		server := &http.Server{
			Addr:      *bind + ":" + *port,
			TLSConfig: tlsConfig,
		}
		fmt.Printf("\nTemporary self signed certificate valid for %v days for the following hostnames:\n%v\n", certValidityDurationDays, sans)
		derBytes := cert.Certificate[0]
		fmt.Printf("\nCertificate fingerprint for checking if the certificate presented by the client's browser is the correct one and not by some MitM attacker:\n%x\n", sha1.Sum(derBytes))

		log.Fatal(server.ListenAndServeTLS("", ""))
	} else {
		log.Fatal(http.ListenAndServe(*bind+":"+*port, nil))
	}
}

// printAddrs prints the local network interfaces and their IP addresses
func printAddrs(scheme, port string) {
	fmt.Println("\nLocal network interfaces and their IP addresses so you can pass one to your colleagues:")
	ifaces, err := net.Interfaces()
	if err != nil {
		log.Fatal(err)
	}
	// We want the interface + IP address list to look like a table with a width of 80:
	//       Interface      |  IPv4 Address   |              IPv6 Address
	// ---------------------|-----------------|----------------------------------------
	// vEthernet (Standan.. | 192.168.178.123 | 1a23:850a:bf55:39a9:6dae:c378:9deb:5aff
	fmt.Println("      Interface      |  IPv4 Address   |              IPv6 Address              ")
	fmt.Println("---------------------|-----------------|----------------------------------------")
	fav := ""
	for _, iface := range ifaces {
		fmt.Printf("%-20v |", cutString(iface.Name, 20))

		// Select IPv4 and IPv6 address
		ipv4, ipv6 := getAddressesFromIface(iface)

		// If there's no favorite IPv4 address yet, check if we should pick the current one
		if fav == "" && isFav(iface) {
			fav = ipv4
		}

		fmt.Printf(" %-15v | %v\n", ipv4, ipv6)
	}

	// Show probable favorite
	if fav != "" {
		fmt.Printf("\nYou probably want to share:\n%v://%v:%v\n", scheme, fav, port)
	}
}
